File: python3/longest_string_chain.py
from typing import List

"""
Space : O(n)
Time  : O(len(word) * max len(word[i]))

Mehtod:
DP - bottom up
"""


class Solution:
    def longestStrChain(self, words: List[str]) -> int:
        ans = 0
        dp = {}
        words.sort(key=lambda w: len(w))
        for word in words:
            for i in range(len(word)):
                mask = word[:i] + word[i+1:]
                dp[word] = max(dp.get(mask, 0) + 1, dp.get(word, 0))
                ans = max(ans, dp[word])

        return ans

File: golang/longest_string_chain.go
package main

import "sort"

// space	: O(n)
// time		: O(ns)

func longestStrChain(words []string) int {
	ans := 0
	dp := map[string]int{}
	sort.Slice(words, func(i, j int) bool {
		return len(words[i]) < len(words[j])
	})

	for _, word := range words {
		dp[word] = 1
		for i := 0; i < len(word); i++ {
			predecessor := word[:i] + word[i+1:]
			if _, ok := dp[predecessor]; ok {
				if dp[word] < dp[predecessor]+1 {
					dp[word] = dp[predecessor] + 1
				}
			}
		}
		ans = Max(ans, dp[word])
	}

	return ans
}
